contracts/EmergencyLiquidator.sol

```diff
@@ -61,66 +80,108 @@ contract EmergencyLiquidator is ACLNonReentrantTrait, IEmergencyLiquidatorExcept
-
-                if (cumulativeLossAfter > cumulativeLossBefore) {
-                    revert NonWhitelistedLiquidationWithLossException();
-                }
-            }
+    /// @dev Checks that either the temporary non-whitelisted mode is enabled, or the msg.sender is whitelised
+    modifier timedNonWhitelistedOnly() {
+        if (block.timestamp > lastWhitelistDisabledTimestamp + whitelistDisabledDuration && !isWhitelisted[msg.sender])
+        {
+            revert CallerNotWhitelistedException();
         }
+        _;
     }
 
     /// @dev Checks that all withdrawals are sent to this contract, reverts if not
     modifier checkWithdrawalDestinations(address creditFacade, MultiCall[] calldata calls) {
-        _checkWithdrawalsDestination(creditFacade, calls);
         _;
     }
 
-    /// @notice Liquidates a credit account, while checking restrictions on liquidations during pause (if any)
-    function liquidateCreditAccount(address creditFacade, address creditAccount, MultiCall[] calldata calls)
+    /// @notice Liquidates a credit account, while checking restrictions on liquidations during pause
+    function liquidateCreditAccount(address creditManager, address creditAccount, MultiCall[] calldata calls)
         external
-        checkWithdrawalDestinations(creditFacade, calls)
-        checkWhitelistedActions(creditFacade)
+        timedNonWhitelistedOnly
     {
-        ICreditFacadeV3(creditFacade).liquidateCreditAccount(creditAccount, address(this), calls);
+        address creditFacade = ICreditManagerV3(creditManager).creditFacade();
+        _checkWithdrawalsDestination(creditFacade, calls);
+        MultiCall[] memory mCalls = _applyPriceFeedUpdates(creditManager, calls);
+
+        CollateralDebtData memory cdd =
+            ICreditManagerV3(creditManager).calcDebtAndCollateral(creditAccount, CollateralCalcTask.DEBT_COLLATERAL);
+        if (
+            _hasBadDebt(creditManager, cdd)
+                && !(_isPolicyWaived(msg.sender) || _isLiquidatableAliased(creditManager, creditAccount, cdd))
+        ) {
+            revert PolicyViolatingLiquidationException();
+        }
+
+        ICreditFacadeV3(creditFacade).liquidateCreditAccount(creditAccount, address(this), mCalls);
     }
 
     /// @notice Liquidates a credit account with max underlying approval, allowing to buy collateral with DAO funds
     /// @dev Can be exploited by account owners when open to everyone, and thus is only allowed for whitelisted addresses
-    function liquidateCreditAccountWithApproval(address creditFacade, address creditAccount, MultiCall[] calldata calls)
-        external
-        checkWithdrawalDestinations(creditFacade, calls)
-        whitelistedOnly
-    {
-        address creditManager = ICreditFacadeV3(creditFacade).creditManager();
+    function liquidateCreditAccountWithApproval(
+        address creditManager,
+        address creditAccount,
+        MultiCall[] calldata calls
+    ) external whitelistedOnly {
```

Comment:
Porbably, it's good to make the contract pausable to block all liquidations if needed